### PR TITLE
(RE-6013, RE-6012, RE-6116, RE-611) Build OpenSSL, Curl, and our ca certs for windows

### PR DESCRIPTION
**Pull request by melissa**
This initial commit adds in the base work to build the puppet-agent project on windows, and modifies the curl, openssl, puppet-ca-bundle and runtime components to run on windows.

### CODE REVIEW

File: configs/projects/puppet-agent.rb [hunk 1]
```diff
@@ -1,34 +1,63 @@
...
 
-  if platform.is_eos?
-    proj.setting(:sysconfdir, "/persist/sys/etc/puppetlabs")
-    proj.setting(:link_sysconfdir, "/etc/puppetlabs")
-  elsif platform.is_osx?
-    proj.setting(:sysconfdir, "/private/etc/puppetlabs")
+  # Project level settings our components will care about
+  # Windows has its own seperate layout
+  if platform.is_windows?
+    proj.setting(:install_root, "#{platform.drive_root}/ProgramFiles/PuppetLabs")
```
Comment (melissa) [inline configs/projects/puppet-agent.rb]: from @joshcooper re: getting around spaces in paths: `"#{platform.drive_root}/Progra~1/Puppet~1"`

Comment (joshcooper) [inline configs/projects/puppet-agent.rb]: Also I should mention that the above isn't 100% reliable, as it could be Progra~2 or Puppet~2. It all depends on what other directories exist. That said if we're building on clean VMs I'm less concerned.

**Comment by melissa** [comment]
32 bit OpenSSL is failing to build :(

File: configs/components/curl.rb [hunk 2]
```diff
@@ -4,18 +4,45 @@
...
+
+  if platform.is_windows?
+    pkg.build_requires "pl-zlib-#{platform.architecture}"
+    pkg.build_requires "runtime"
+
+    arch = platform.architecture == "x64" ? "64" : "32"
+    pkg.environment "PATH" => "#{settings[:gcc_bindir]}:$$PATH"
+    pkg.environment "CYGWIN" => settings[:cygwin]
+    pkg.environment "CC" => settings[:cc]
+    pkg.environment "CXX" => settings[:cxx]
```
Comment (jcoconnor) [inline configs/components/curl.rb]: Should we adopt these "settings" for pl-build-tools as well?

**Comment by jcoconnor** [comment]
Background question - why aren't be building this in pl-build-tools ?

File: configs/components/curl.rb [hunk 3]
```diff
@@ -4,18 +4,45 @@
...
   end
 
   pkg.build do
-    ["#{platform[:make]} -j$(shell expr $(shell #{platform[:num_cores]}) + 1)"]
+    ["#{make} -j$(shell expr $(shell #{platform[:num_cores]}) + 1)"]
   end
 
   pkg.install do
-    ["#{platform[:make]} -j$(shell expr $(shell #{platform[:num_cores]}) + 1) install"]
+    ["#{make} -j$(shell expr $(shell #{platform[:num_cores]}) + 1) install"]
```
Comment (jcoconnor) [inline configs/components/curl.rb]: Why the change and should we do this in pl-tools-vanagon ?

File: configs/projects/puppet-agent.rb (hunk 1 above)
Comment (glennsarti) [inline configs/projects/puppet-agent.rb]: @joshcooper +1 to you're pretty safe with shortnames for Program Files.  Another gotcha will be ProgFiles vs ProgFiles (x86).  Would this be something to be concerned about?

File: configs/projects/puppet-agent.rb [hunk 4]
```diff
@@ -47,6 +76,14 @@
         proj.setting(:host_gem, "/opt/pl-build-tools/bin/gem")
       end
     end
+  elsif platform.is_windows?
+    # For windows, we need to ensure we are building for mingw not cygwin
+    if platform.architecture == "x64"
+      platform_triple = "x86_64-unknown-mingw32"
+    else
+      platform_triple = "i686-unknown-mingw32"
+    end
```
Comment (jcoconnor) [inline configs/projects/puppet-agent.rb]: Use Ternary here ?

File: configs/components/openssl.rb [hunk 5]
```diff
@@ -54,6 +66,23 @@
...
     cflags = "$${CFLAGS} -static-libgcc"
     pkg.environment "CC" => "/opt/pl-build-tools/bin/gcc"
+  elsif platform.is_windows?
+    if platform.architecture == "x64"
+      target = "mingw64"
+      arch = "64"
+    else
+      target = "mingw"
+      arch = "32"
+    end
```
Comment (jcoconnor) [inline configs/components/openssl.rb]: Why not use ternary here similar to runtime.rb below?

File: configs/components/curl.rb (hunk 2 above)
Comment (melissa) [inline configs/components/curl.rb]: If you want to... I don't want to go back and rebuild all those packages just for this though

File: configs/components/curl.rb (hunk 3 above)
Comment (melissa) [inline configs/components/curl.rb]: Because we're calling a different make... Is pl-build-tools-vanagon not working?

File: configs/projects/puppet-agent.rb (hunk 1 above)
Comment (melissa) [inline configs/projects/puppet-agent.rb]: @glennsarti I think we're just installing in ProgFiles, so we shouldn't have to worry about ProgFiles (x86). Josh can verify that though

File: configs/components/curl.rb (hunk 3 above)
Comment (jcoconnor) [inline configs/components/curl.rb]: ok - was just asking ;)

File: configs/projects/puppet-agent.rb (hunk 1 above)
Comment (joshcooper) [inline configs/projects/puppet-agent.rb]: @glennsarti good point. I was assuming we're always building 32-bit MSIs on 32-bit Windows and building 64-bit MSIs on 64-bit Windows, but I think that's a bad assumption, since I think we always build on 2012r2. So yeah ProgFiles (x86) vs ProgFiles could be an issue. It appears that ~1 maps to ProgFiles and ~2 maps to ProgFiles (x86):

```
Administrator@mm1fxccjkyk8b1h /cygdrive/c
$ cygpath -ad "Program Files"
C:\PROGRA~1

Administrator@mm1fxccjkyk8b1h /cygdrive/c
$ cygpath -ad "Program Files (x86)"
C:\PROGRA~2
```

I would expect that to always be the case, but maybe not?

Comment (glennsarti) [inline configs/projects/puppet-agent.rb]: The program files dirs are laid down during OS install so while it is _possible_ to make them not ~1 and ~2, it's such an extreme edge case.  Also it is possible to disable 8.3 name generation in NTFS but again, it's an edge case and _generally_ you wouldn't do it prior to the OS installation there fore the 8.3 names for prog file will still get generated.

File: configs/components/curl.rb (hunk 3 above)
Comment (melissa) [inline configs/components/curl.rb]: if you want to make this change in pl-build-tools-vanagon, you're welcome to, but I'm really not worried about it. I consider that repo our practice ground. We found out some things that worked and some things that didn't. In puppet-agent, we get to make things more elegant and official.

File: configs/components/openssl.rb (hunk 5 above)
Comment (melissa) [inline configs/components/openssl.rb]: No reason. It just seemed more legible at the time. Do you think it would be more logical to have these be ternaries?

Comment (jcoconnor) [inline configs/components/openssl.rb]: More consistency really - functionally doesn't make a difference. You definitely don't need a ternary on line 116+ as it would make it unreadable.

**Comment by jcoconnor** [comment]
Have successfully built this for Windows 64 bit, with the following mods to the build env:
* Disable facter, ruby depends (not done yet).
* Ensure latest vanagon/vanagon_for_the_win branch is pulled.
* Cherry picked in change 74b44ad34ec9bce671ca530b2113b9aa2e912022 to vanagon.

So this looks good (on 64).
:+1:

**Comment by melissa** [comment]
@jcoconnor what are you asking about above when you ask why we're not building it in pl-build-tools? OpenSSL? Curl? Puppet ca bundle? Runtime? We're building a lot of things in this PR... You have to be a little more specific with your question.

The main reason is that we need to deliver the files as a part of the puppet agent package. The easiest way to do that is build them here so vanagon knows to include the files in the final package (remember the file-list functionality I was telling you about?).

File: configs/projects/puppet-agent.rb (hunk 1 above)
Comment (melissa) [inline configs/projects/puppet-agent.rb]: @joshcooper how do I go about setting it up to pick up `Puppet~1` as Puppet Labs? Is that just going to look for the first directory in that path with a space in it?

Comment (joshcooper) [inline configs/projects/puppet-agent.rb]: I think it's safe to hard code as Puppet~1 since the VM shouldn't have Puppet already installed.

File: configs/components/curl.rb [hunk 6]
```diff
@@ -4,18 +4,45 @@
...
+  prefix = settings[:prefix]
+  pem_file = "#{settings[:prefix]}/ssl/cert.pem"
+
+  make = platform[:make]
+
+  if platform.is_windows?
+    pkg.build_requires "pl-zlib-#{platform.architecture}"
+    pkg.build_requires "runtime"
+
+    arch = platform.architecture == "x64" ? "64" : "32"
```
Comment (MikaelSmith) [inline configs/components/curl.rb]: This looks like it's no longer needed.

File: configs/components/openssl.rb [hunk 7]
```diff
@@ -54,6 +66,23 @@
     ldflags = ''
     cflags = "$${CFLAGS} -static-libgcc"
     pkg.environment "CC" => "/opt/pl-build-tools/bin/gcc"
+  elsif platform.is_windows?
+    if platform.architecture == "x64"
```
Comment (MikaelSmith) [inline configs/components/openssl.rb]: arch is no longer needed, this could be replaced with `target = platform.architecture == "x64" ? "mingw64" : "mingw"`

File: configs/components/openssl.rb [hunk 8]
```diff
@@ -54,6 +66,23 @@
...
+      target = "mingw64"
+      arch = "64"
+    else
+      target = "mingw"
+      arch = "32"
+    end
+    pkg.environment "PATH" => "#{settings[:gcc_bindir]}:#{settings[:tools_root]}/bin:$$PATH"
```
Comment (MikaelSmith) [inline configs/components/openssl.rb]: Why tools_root here?

File: configs/components/runtime.rb [hunk 9]
```diff
@@ -23,6 +25,16 @@
   if platform.is_aix?
     pkg.install_file File.join(libdir, "libstdc++.a"), "/opt/puppetlabs/puppet/lib/libstdc++.a"
     pkg.install_file File.join(libdir, "libgcc_s.a"), "/opt/puppetlabs/puppet/lib/libgcc_s.a"
+  elsif platform.is_windows?
+    arch = platform.architecture == "x64" ? "64" : "32"
+    lib_type = platform.architecture == "x64" ? "seh" : "sjlj"
+    pkg.install_file "#{platform.drive_root}/tools/mingw#{arch}/bin/libgcc_s_#{lib_type}-1.dll", "#{settings[:bindir]}/libgcc_s_#{lib_type}-1.dll"
```
Comment (MikaelSmith) [inline configs/components/runtime.rb]: This could use gcc_bindir.

File: configs/projects/puppet-agent.rb [hunk 10]
```diff
@@ -1,34 +1,68 @@
...
-  elsif platform.is_osx?
-    proj.setting(:sysconfdir, "/private/etc/puppetlabs")
+  # Project level settings our components will care about
+  # Windows has its own seperate layout
+  if platform.is_windows?
+    # Our install prefix can't have spaces in it, so we take advantage of short cuts.
+    # However, in order to take advantage of these shortcuts, we need to explicitely
+    # create the directories.
+    proj.setting(:install_root, "#{platform.drive_root}/Progra~1/Puppet~1")
+    proj.directory "#{platform.drive_root}/Program Files/Puppet Labs"
```
Comment (melissa) [inline configs/projects/puppet-agent.rb]: @joshcooper I'm pretty sure this is working now! The only issue is that the install prefix is changing from `Puppet` to `puppet`, and I'm not sure why. We're setting it to upper case puppet on line 13.

Comment (MikaelSmith) [inline configs/projects/puppet-agent.rb]: Down below directory creation is using install_root. I'm not sure what that will do...

Comment (melissa) [inline configs/projects/puppet-agent.rb]: Good point, I'm only setting that if we're not building on windows

File: configs/projects/puppet-agent.rb (hunk 4 above)
Comment (melissa) [inline configs/projects/puppet-agent.rb]: I think that would be too long of a line. This is more readable.

File: configs/components/curl.rb [hunk 11]
```diff
@@ -4,18 +4,44 @@
...
+    pkg.build_requires "pl-zlib-#{platform.architecture}"
+    pkg.build_requires "runtime"
+
+    pkg.environment "PATH" => "#{settings[:gcc_bindir]}:$$PATH"
+    pkg.environment "CYGWIN" => settings[:cygwin]
+    pkg.environment "CC" => settings[:cc]
+    pkg.environment "CXX" => settings[:cxx]
+
+    prefix = platform.convert_to_windows_path(settings[:prefix])
+    pem_file = platform.convert_to_windows_path("#{settings[:puppet_configdir]}/ssl/cert.pem")
```
Comment (melissa) [inline configs/components/curl.rb]: @joshcooper is this where we decided to install the pem file? This currently expands to `C:/ProgramData/PuppetLabs/puppet/etc/ssl/cert.pem`

Comment (joshcooper) [inline configs/components/curl.rb]: The ssldir looks correct based on https://github.com/puppetlabs/puppet-specifications/blame/master/file_paths.md#L222. But is it safe to drop `cert.pem` in there? Or can it conflict with things puppet may write to its `ssldir`?

Comment (MikaelSmith) [inline configs/components/curl.rb]: I would probably give it a more descriptive name.

However, I don't see anything it would conflict with. Nothing at https://github.com/puppetlabs/puppet/blob/master/lib/puppet/defaults.rb#L645-L770 does.

File: configs/projects/puppet-agent.rb [hunk 12]
```diff
@@ -82,8 +121,26 @@
   end
 
   # Platform specific
-  proj.setting(:cflags, "-I#{proj.includedir} -I/opt/pl-build-tools/include")
-  proj.setting(:ldflags, "-L#{proj.libdir} -L/opt/pl-build-tools/lib -Wl,-rpath=#{proj.libdir}")
+  if platform.is_windows?
+    if platform.architecture == "x64"
+      arch = "64"
+      proj.setting(:cflags, "-IC:/tools/pl-build-tools/include -IC:/tools/mingw#{arch}/include -IC:/tools/mingw64/x86_64-w64-mingw32/include -I#{platform.convert_to_windows_path(proj.includedir)}")
```
Comment (MikaelSmith) [inline configs/projects/puppet-agent.rb]: Why do we need mingw64/x86_64-w64/mingw32/include? GCC should already know about that location.

**Comment by MikaelSmith** [comment]
Other than what seems like extra settings in cflags and ldflags, :+1:

File: configs/components/curl.rb [hunk 13]
```diff
@@ -4,18 +4,44 @@
...
+
+    make = "/usr/bin/make"
```
Comment (stahnma) [inline configs/components/curl.rb]: Is there a reason we aren't using platform make throughout?

File: configs/components/curl.rb [hunk 14]
```diff
@@ -4,18 +4,44 @@
...
+        --enable-threaded-resolver \
+        --disable-ldap \
+        --disable-ldaps \
+        --with-ca-bundle=#{pem_file} \
+        #{settings[:host]}"]
   end
 
   pkg.build do
-    ["#{platform[:make]} -j$(shell expr $(shell #{platform[:num_cores]}) + 1)"]
+    ["#{make} -j$(shell expr $(shell #{platform[:num_cores]}) + 1)"]
```
Comment (stahnma) [inline configs/components/curl.rb]: Why change from platform make to make?

File: configs/components/curl.rb (hunk 3 above)
Comment (stahnma) [inline configs/components/curl.rb]: Hmm. I think if we're going to call a different make here, we should make the default still be platform[:make] rather than make = "/usr/bin/make"...

File: configs/components/curl.rb (hunk 13 above)
Comment (stahnma) [inline configs/components/curl.rb]: Nevermind, I get it.

File: configs/components/curl.rb (hunk 14 above)
Comment (stahnma) [inline configs/components/curl.rb]: Missed the reassignment above. :fearful:

File: configs/components/curl.rb (hunk 3 above)
Comment (stahnma) [inline configs/components/curl.rb]: You did that, I misunderstood what was happening.

**Comment by stahnma** [comment]
:+1:

File: configs/components/curl.rb (hunk 3 above)
Comment (haus) [inline configs/components/curl.rb]: If a different make is being used for all components in the agent, it would be worth setting `platform.make "/usr/bin/make"` in the windows platform config. Then you can avoid the intermediate variable here.

File: configs/projects/puppet-agent.rb (hunk 4 above)
Comment (haus) [inline configs/projects/puppet-agent.rb]: This is settable in the platform config.

File: configs/platforms/win-2012-x64.rb [hunk 15]
```diff
@@ -0,0 +1,19 @@
+platform "win-2012-x64" do |plat|
+  plat.vmpooler_template "win-2012r2-x86_64.make"
+
```
Comment (haus) [inline configs/platforms/win-2012-x64.rb]: you can define the platform triple here and in the x86 one below instead of in the project config.

File: configs/components/curl.rb (hunk 3 above)
Comment (melissa) [inline configs/components/curl.rb]: @haus it's not being used by all components :(

Comment (haus) [inline configs/components/curl.rb]: sadness

Comment (jcoconnor) [inline configs/components/curl.rb]: sadness is right, but we have found a small subset of components that won't work with ming-make, hence the need to change it. libffi in pl-build-tools-vanagon is another example that needs MSYS make or equiv.

File: configs/components/openssl.rb [hunk 16]
```diff
@@ -54,6 +66,17 @@
     ldflags = ''
     cflags = "$${CFLAGS} -static-libgcc"
     pkg.environment "CC" => "/opt/pl-build-tools/bin/gcc"
+  elsif platform.is_windows?
+    target = platform.architecture == "x64" ? "mingw64" : "mingw"
+    pkg.environment "PATH" => "#{settings[:gcc_bindir]}:$$PATH"
+    pkg.environment "CYGWIN" => settings[:cygwin]
+    pkg.environment "CC" => settings[:cc]
+    pkg.environment "CXX" => settings[:cxx]
+    make = "/usr/bin/make"
```
Comment (jcoconnor) [inline configs/components/openssl.rb]: Minor issue - my preference would be to group all the pkg.environment statements together and put the make reassignment after them. This is an odd-case, but wondering should you put the make assignment back to line 70 and follow it with pkg.environment for the make (if you think this is OCD speaking, feel free to ignore).

File: configs/platforms/win-2012-x64.rb (hunk 15 above)
Comment (melissa) [inline configs/platforms/win-2012-x64.rb]: done
